### PR TITLE
bpo-35995: update logging.handlers.SMTPHandler to support smtp server which just in TLS mode

### PR DESCRIPTION
**Pull request by lidayan**
[bpo-35995](https://bugs.python.org/issue35995): Fix some bug in logging module

just setting param secure=[] means use ssl socket with default certfile
but smtplib.SMTP() failed in self.connect --> self._get_socket. because it is not SMTP_SSL._get_socket

https://bugs.python.org/issue35995

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/logging/handlers.py
Comment (lidayan) [inline Lib/logging/handlers.py:1059]: format code style by hande

**Review by eamanu — CHANGES_REQUESTED** [review]
Please, could you add some test? Also, I think that this need have a news. 

And please sign CLA

**Review by JulienPalard — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution!

Looks like you're changing from explicit TLS (STARTSSL) to implicit TLS, this may fix your communication with your server, but this will break the communcation to servers using explicit TLS.

**Comment by lidayan** [comment]
> A Python core developer has requested some changes be made to your pull request before we can consider merging it. If you could please address their requests along with any other requests in other reviews from core developers that would be appreciated.
> 
> Once you have made the requested changes, please leave a comment on this pull request containing the phrase `I have made the requested changes; please review again`. I will then notify any core developers who have left a review that you're ready for them to take another look at this pull request.
> 
> And if you don't make the requested changes, [you will be put in the comfy chair!](https://youtu.be/XnS49c9KZw8?t=1m30s)

I got it

**Comment by lidayan** [comment]
> Please, could you add some test? Also, I think that this need have a news.
> 
> And please sign CLA

yes, I have tested it in my project

**Comment by lidayan** [comment]
> Looks like you're changing from explicit TLS (STARTSSL) to implicit TLS, this may fix your communication with your server, but this will break the communcation to servers using explicit TLS.

use "ssl._create_stdlib_context" and "SMTP_SSL._get_socket" when init smtp, just like startssl function.

**Comment by lidayan** [comment]
> Looks like you're changing from explicit TLS (STARTSSL) to implicit TLS, this may fix your communication with your server, but this will break the communcation to servers using explicit TLS.

What's your means between "explicit TLS" and "implicit TLS"?

when set usename and secure(not none) just use SMTP_SSL instead of SMTP. the conditions is not changed

**Comment by matrixise** [comment]
Hi @lidayan, @JulienPalard 

What are the next steps for this PR?

**Comment by lidayan** [comment]
> Hi @lidayan, @JulienPalard
> 
> What are the next steps for this PR?

if no other problems, I wish it can be accepted to master branch.

**Comment by lidayan** [comment]
@JulienPalard

Two solutions for this issue,  update SMTPHandler or add New SMTPSSLHandler, I not sure which is better

**Comment by vsajip** [comment]
I propose to close this as not needed in the stdlib. Also, this is not a bug in logging, so the title is not correct, it seems to me.

**Comment by lidayan** [comment]
@vsajip

using logging.handler.SMTPHandler with secure=[], but the smtp server just support tls mode, then it's error. this case need not support?

**Comment by lidayan** [comment]
@vsajip

using logging.handler.SMTPHandler with secure=[], but the smtp server just support tls mode, then it's error. this case need not support?

**Comment by lidayan** [comment]
@vsajip

using logging.handler.SMTPHandler with secure=[], but the smtp server just support tls mode, then it's error. this case need not support?

**Comment by lidayan** [comment]
@vsajip

using logging.handler.SMTPHandler with secure=[], but the smtp server just support tls mode, then it's error. this case need not support?

**Comment by lidayan** [comment]
I propose to update SMTHHandler to support the smtp server whitch just in TLS mode.

**Comment by lidayan** [comment]
I propose to update SMTHHandler to support the smtp server whitch just in TLS mode. because it owns argument "secure" and allow secure=[]

**Comment by matrixise** [comment]
Do you close it? or do you open it? if you open it, please change the title of your PR, because "Fix some bug in logging module" is really ambiguous :/

Thank you

**Comment by vsajip** [comment]
You have removed the STARTTLS stuff - why have you done that? Won't it affect backwards compatibility? Not all servers will be working like the one you're communicating with, will they? For the many years that SMTPHandler has been around (since the very beginning - 2003) there has not been a problem with the support using STARTTLS as far as I know.

Also, there is no test, nor any update to the documentation, so IMO this PR is not complete. I suggest to close this PR unless another core developer makes a case for this functionality to be in the stdlib right now.

**Comment by lidayan** [comment]
@vsajip 
 with argument secure=[] and SMTPHandler initialize self.smtp=smtplib.SMTP(), right here in SMTP init method calling smtp.connect and smtp._get_socket as plaintext mode. It is timeout error.  
I think using "secure=[]" means connecting as TLS mode

**Comment by lidayan** [comment]
@vsajip 
It base on the original intention for SMTPHandler, and argument "secure=[]" means what?

**Comment by lidayan** [comment]
I respect your operations,  after all python is a magnificent project, I like it and wish it better.

**Comment by vsajip** [comment]
> I think using "secure=[]" means connecting as TLS mode
...
> It base on the original intention for SMTPHandler, and argument "secure=[]" means what?

The meaning of `secure` is reasonably well documented:

> To specify the use of a secure protocol (TLS), pass in a tuple to the `secure` argument. This will only be used when authentication credentials are supplied. The tuple should be either an empty tuple, or a single-value tuple with the name of a keyfile, or a 2-value tuple with the names of the keyfile and certificate file. (This tuple is passed to the `smtplib.SMTP.starttls()` method.)

Any change to this functionality must:

* Not break backwards compatibility.
* Be accompanied by changes to tests.
* Be accompanied by documentation updates.

The current situation (out there in general, not with `SMTPHandler`) is a bit of a mess. From [this source](https://www.fastmail.com/help/technical/ssltlsstarttls.html):

> The result is that in most cases, systems that offer message submission over port 587 require clients to use STARTTLS to upgrade the connection and also require a username and password to authenticate. There has been an added benefit to this approach as well. By moving users away from using port 25 for email submission, ISPs are now able to block outgoing port 25 connections from users' computers, which were a significant source of spam due to infection with spam-sending viruses.

> Unfortunately the downside of changing port numbers is that a number of email clients were made which only supported implicit SSL/TLS over port 465 and not STARTTLS on 587. Clients are often very long lived, and so removing port 465 wasn't an option for many sites without annoying customers. Additionally, because port 465 was advertised as an option, many users with email clients that support both STARTTLS on 587 and SSL/TLS on 465 set them up to use 465 instead of 587. This makes it even harder to remove support for port 465, since lots of users have their email clients set up to use it.

> Currently, things seem relatively randomly split between people using SMTP SSL/TLS encrypted over port 465, and people using SMTP with STARTTLS upgrading over port 587.

> And just to add extra confusion, in 2018 it was decided to change yet again, and recommend using implicit TLS over port 465 to "encourage more widespread use of TLS and to also encourage greater consistency regarding how TLS is used, this specification now recommends the use of Implicit TLS for POP, IMAP, SMTP Submission, and all other protocols used between an MUA and an MSP."

Note that the recommendation to use implicit TLS is relatively new, and most servers will not have been upgraded. Therefore, the existing STARTTLS functionality must be kept exactly as it, for the time being.

I have no objection to a patch which provides implicit TLS support, but this must be an opt-in thing for individual uses of `SMTPHandler`, and not a blanket change which replaces STARTTLS. Also, test and documentation changes are needed (including a test which fires up a suitable server component to test implicit TLS). All of these things are _not_ featured in this PR.

> I respect your operations, after all python is a magnificent project, I like it and wish it better.

Sure, and thanks for trying to improve things. I just want to make sure that any patch meets the conditions described above, before it can be seriously considered for merging.

**Comment by vsajip** [comment]
Closed as per my comments above. I've also closed the corresponding issue. Feel free to reopen PR and issue if you come up with a patch conforming to the guidelines I posted.

**Pull request closed without merging**
